simulator/virtual_machine.go

```diff
@@ -639,3 +723,54 @@ func (vm *VirtualMachine) ShutdownGuest(c *types.ShutdownGuest) soap.HasFault {
 
 	return r
 }
+
+func findSnapshotInTree(tree []types.VirtualMachineSnapshotTree, ref types.ManagedObjectReference) *types.VirtualMachineSnapshotTree {
+	if tree == nil {
+		return nil
+	}
+
+	for i, ss := range tree {
+		if ss.Snapshot == ref {
+			return &tree[i]
+		}
+
+		res := findSnapshotInTree(ss.ChildSnapshotList, ref)
+		if res != nil {
+			return res
+		}
+	}
+	return nil
+}
+
+func removeSnapshotInTree(tree []types.VirtualMachineSnapshotTree, ref types.ManagedObjectReference, removeChildren bool) []types.VirtualMachineSnapshotTree {
+	if tree == nil {
+		return tree
+	}
+
+	var result []types.VirtualMachineSnapshotTree
+	for _, ss := range tree {
+		if ss.Snapshot == ref {
+			if !removeChildren {
+				result = append(result, ss.ChildSnapshotList...)
+			}
+		} else {
+			ss.ChildSnapshotList = removeSnapshotInTree(ss.ChildSnapshotList, ref, removeChildren)
+			result = append(result, ss)
+		}
+	}
+	return result
+}
+
+func allSnapshotsInTree(tree []types.VirtualMachineSnapshotTree) []types.ManagedObjectReference {
+	var result []types.ManagedObjectReference
+
+	if tree == nil {
+		return result
+	}
+
+	for _, ss := range tree {
+		result = append(result, ss.Snapshot)
+		result = append(result, allSnapshotsInTree(ss.ChildSnapshotList)...)
+	}
+	return result
```

Comment:
style: add an empty line between the statements (there's a few of these in the PR), we do that in most cases for return too:
`git grep -A1 -B2 'return ' -- simulator`